packages/@storefront/flux-capacitor/src/core/emitter.ts

```diff
@@ -0,0 +1,90 @@
+import { EventEmitter } from 'eventemitter3';
+
+class Emitter extends EventEmitter {
+  _barriers: Barriers;
+  _lookups: Lookups;
+
+  constructor() {
+    super();
+    this._barriers = {};
+    this._lookups = {};
+  }
+
+  emit(event: string, ...args: any[]): boolean {
+    const result = super.emit(event, ...args);
+    const keys = this._lookups[event] || [];
+
+    keys.forEach((key) => {
+      this._barriers[key].events[event]++;
+
+      const shouldInvoke = Object.keys(this._barriers[key].events).every((ev) => this._barriers[key].events[ev] > 0);
+
+      if (shouldInvoke) {
+        this._barriers[key].callbacks.forEach(({ callback, context }) => callback.apply(context));
+        // tslint:disable-next-line max-line-length
+        this._barriers[key].events = Object.keys(this._barriers[key].events).reduce((acc, ev) => Object.assign(acc, { [ev]: 0 }), {});
+      }
+    });
+
+    return result;
+  }
+
+  all(events: string[], callback: () => void, context: any = this) {
+    const key = this.generateKey(events);
+
+    this._barriers[key] = {
+      events: events.reduce((acc, ev) => ({ ...acc, [ev]: 0 }), {}),
+      callbacks: this._barriers[key]
+        ? [...this._barriers[key].callbacks, { callback, context }]
+        : [{ callback, context }],
+    };
+
+    // tslint:disable-next-line max-line-length
+    this._lookups = events.reduce((acc, ev) => Object.assign(acc, { [ev]: this._lookups[ev] ? [...this._lookups[ev], key] : [key] }), this._lookups);
```

Comment:
@jmbtutor asked for `Object.assign()`. Consensus?

Reply:
We do this `Object.assign` pattern elsewhere in the code, and using `Object.assign` means that the amount of garbage that you create with the intermediary objects is just O(n) instead of O(n<sup>2</sup>). Ideally, we create no garbage by doing a for-each and assigning directly.

On that note, re-assigning to `this._lookups` is redundant because you're assigning the object to itself, since `this._lookups` is used as the initial accumulator.

Reply:
mb didn't realize he asked you to switch to it. It's not worth arguing over.

Reply:
Leaving as-is, marking as resolved.